linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala

```diff
@@ -1316,139 +1316,120 @@ private final class Analyzer(config: CommonPhaseConfig,
         staticDependencies += info.className
     }
 
-    for (moduleName <- data.accessedModules) {
-      lookupClass(moduleName) { module =>
-        module.accessModule()
-        addInstanceDependency(module)
-      }
-    }
+    for (dataInClass <- data.byClass) {
+      lookupClass(dataInClass.className) { clazz =>
+        val className = dataInClass.className
 
-    for (className <- data.instantiatedClasses) {
-      lookupClass(className) { clazz =>
-        clazz.instantiated()
-        addInstanceDependency(clazz)
-      }
-    }
+        val flags = dataInClass.flags
+        if (flags != 0) {
+          if ((flags & ReachabilityInfoInClass.FlagModuleAccessed) != 0) {
+            clazz.accessModule()
+            addInstanceDependency(clazz)
+          }
 
-    for (className <- data.usedInstanceTests) {
-      staticDependencies += className
-      lookupClass(className)(_.useInstanceTests())
-    }
+          if ((flags & ReachabilityInfoInClass.FlagInstantiated) != 0) {
+            clazz.instantiated()
+            addInstanceDependency(clazz)
+          }
 
-    for (className <- data.accessedClassData) {
-      staticDependencies += className
-      lookupClass(className)(_.accessData())
-    }
+          if ((flags & ReachabilityInfoInClass.FlagInstanceTestsUsed) != 0) {
+            staticDependencies += className
+            clazz.useInstanceTests()
+          }
 
-    if (data.accessedClassClass) {
-      /* java.lang.Class is only ever instantiated in the CoreJSLib.
-       * Therefore, make java.lang.Object depend on it instead of the caller itself.
-       */
-      objectClassInfo.staticDependencies += ClassClass
-      lookupClass(ClassClass) { clazz =>
-        clazz.instantiated()
-        clazz.callMethodStatically(MemberNamespace.Constructor, ObjectArgConstructorName)
-      }
-    }
+          if ((flags & ReachabilityInfoInClass.FlagClassDataAccessed) != 0) {
+            staticDependencies += className
+            clazz.accessData()
+          }
 
-    for (className <- data.referencedClasses) {
-      /* No need to add to staticDependencies: The classes will not be
-       * referenced in the final JS code.
-       */
-      lookupClass(className)(_ => ())
-    }
+          if ((flags & ReachabilityInfoInClass.FlagStaticallyReferenced) != 0) {
+            staticDependencies += className
+          }
+        }
 
-    for (className <- data.staticallyReferencedClasses) {
-      staticDependencies += className
-      lookupClass(className)(_ => ())
-    }
+        /* Since many of the lists below are likely to be empty, we always
+         * test `!list.isEmpty` before calling `foreach` or any other
+         * processing, avoiding closure allocations.
+         */
```

Comment:
Does that make sense to do it here for the calls that allocate closures further down? (e.g. `readFields`)? Feels like we should adjust the definition sites instead?

Reply:
Meh. It's a trade-off. It only applies to `readFields` and `writeFields`, really. If I move the `isEmpty` check to them, and I read this piece of code, I wonder why those *don't* have the `isEmpty` check. It seems more systematic to do it for every list here, and not wonder.